File: probes-js/langgraph_probes.mjs
import {
  Annotation,
  Command,
  END,
  interrupt,
  MemorySaver,
  START,
  StateGraph,
} from "@langchain/langgraph";

const sleep = (ms) => new Promise((r) => setTimeout(r, ms));

class EventLog {
  constructor() {
    this.events = [];
  }

  log(e) {
    this.events.push(e);
  }

  clear() {
    this.events = [];
  }

  contains(sub) {
    return this.events.some((e) => e.includes(sub));
  }

  count(sub) {
    return this.events.filter((e) => e.includes(sub)).length;
  }
}

const LOG = new EventLog();

const results = [];

function record(name, violation, detail) {
  results.push({ name, violation, detail });
  const tag = violation ? "VIOLATION" : "clean/contrast";
  console.log(`${name.padEnd(32)} -> ${tag}  ${JSON.stringify(detail)}`);
}

const State = Annotation.Root({
  note: Annotation({
    reducer: (_prev, next) => next,
    default: () => "",
  }),
});

async function runJ1() {
  LOG.clear();

  const graph = new StateGraph(State)
    .addNode("gate", async () => {
      LOG.log("gate:before_interrupt");
      const decision = interrupt({ question: "approve?" });
      LOG.log(`gate:resumed_with=${JSON.stringify(decision)}`);
      return { note: "gated" };
    })
    .addNode("side_effect", async () => {
      LOG.log("EFFECT:executed"); // stands in for an email send
      return { note: "effect" };
    })
    .addEdge(START, "gate")
    .addEdge(START, "side_effect")
    .addEdge("gate", END)
    .addEdge("side_effect", END)
    .compile({ checkpointer: new MemorySaver() });

  const cfg = { configurable: { thread_id: "j1" } };
  const first = await graph.invoke({ note: "" }, cfg);
  const paused = Boolean(first.__interrupt__ && first.__interrupt__.length);
  const effectWhilePaused = LOG.contains("EFFECT:executed");
  const traceAtPause = [...LOG.events];

  record("sibling_leak[fan_out]", paused && effectWhilePaused, {
    paused,
    effect_while_paused: effectWhilePaused,
    trace_at_pause: traceAtPause,
  });

  await graph.invoke(new Command({ resume: "reject" }), cfg);
  const effectTotal = LOG.count("EFFECT:executed");

  record("reject_after_effect", paused && effectTotal >= 1, {
    effect_total_after_reject: effectTotal,
  });
}


async function runJ2() {
  LOG.clear();

  const graph = new StateGraph(State)
    .addNode("act_then_gate", async () => {
      LOG.log("J2_EFFECT:executed");
      const decision = interrupt({ question: "approve?" });
      LOG.log(`gate:resumed_with=${JSON.stringify(decision)}`);
      return { note: "done" };
    })
    .addEdge(START, "act_then_gate")
    .addEdge("act_then_gate", END)
    .compile({ checkpointer: new MemorySaver() });

  const cfg = { configurable: { thread_id: "j2" } };
  const first = await graph.invoke({ note: "" }, cfg);
  const paused = Boolean(first.__interrupt__ && first.__interrupt__.length);

  if (!paused) throw new Error("J2 harness failure: interrupt() did not pause");

  const countAtPause = LOG.count("J2_EFFECT:executed");

  await graph.invoke(new Command({ resume: "approve" }), cfg); // human APPROVES

  const countAfter = LOG.count("J2_EFFECT:executed");

  record("replay[resume_after_approve]", countAfter > 1, {
    paused,
    effect_count_at_pause: countAtPause,
    effect_count_after_approve: countAfter,
  });
}

async function runJ3() {
  LOG.clear();
  const graph = new StateGraph(State)
    .addNode("worker", async () => {
      LOG.log("J3:node_started");
      await sleep(600);
      LOG.log("J3_EFFECT:executed_after_delay");
      return { note: "done" };
    })
    .addEdge(START, "worker")
    .addEdge("worker", END)
    .compile({ checkpointer: new MemorySaver() });

  const controller = new AbortController();
  const cfg = { configurable: { thread_id: "j3" }, signal: controller.signal };
  const run = graph.invoke({ note: "" }, cfg);
  await sleep(150); // user cancels mid-node
  controller.abort();
  let abortSeen = false;
  let callerSaw = null;

  try {
    await run;
  } catch (e) {
    callerSaw = e && e.name ? e.name : String(e);
    abortSeen = true;
  }

  const effectAtCancel = LOG.contains("J3_EFFECT");

  await sleep(800);
  const effectAfter = LOG.contains("J3_EFFECT");

  record(
    "cancellation[abort_signal]",
    abortSeen && !effectAtCancel && effectAfter,
    {
      abort_seen: abortSeen,
      caller_saw: callerSaw,
      effect_at_cancel: effectAtCancel,
      effect_after_cancel: effectAfter,
      note: "no sync/async split exists in JS: promises are not cancellable",
    }
  );
}

async function runJ4() {
  LOG.clear();

  const graph = new StateGraph(State)
    .addNode("worker", async () => {
      LOG.log("J4:node_started");
      await sleep(800);
      LOG.log("J4_EFFECT:executed_after_delay");
      return { note: "done" };
    })
    .addEdge(START, "worker")
    .addEdge("worker", END)
    .compile({ checkpointer: new MemorySaver() });

  const cfg = {
    configurable: { thread_id: "j4" },
    signal: AbortSignal.timeout(200),
  };

  let timedOut = false;

  let callerSaw = null;

  try {
    await graph.invoke({ note: "" }, cfg);
  } catch (e) {
    callerSaw = e && e.name ? e.name : String(e);
    timedOut = true;
  }

  const effectAtTimeout = LOG.contains("J4_EFFECT");

  await sleep(1000);

  const effectAfter = LOG.contains("J4_EFFECT");

  record(
    "timeout_zombie[abort_timeout]",
    timedOut && !effectAtTimeout && effectAfter,
    {
      caller_saw: callerSaw,
      effect_at_timeout: effectAtTimeout,
      effect_after_timeout: effectAfter,
    }
  );
}

async function main() {
  const pkg = (await import("./node_modules/@langchain/langgraph/package.json", {
    with: { type: "json" },
  })).default;

  console.log(`# FW-F @langchain/langgraph@${pkg.version} (Node ${process.version})\n`);

  await runJ1();
  await runJ2();
  await runJ3();
  await runJ4();
  const v = results.filter((r) => r.violation).length;

  console.log(`\nVIOLATION AXES CONFIRMED: ${v}/${results.length}`);
}

main().catch((e) => {
  console.error("PROBE HARNESS FAILURE:", e);
  process.exit(1);
});

File: probes/src/agentprobe/langgraph_probes.py
from __future__ import annotations
import asyncio
import time
from importlib.metadata import version
from typing import TypedDict

from langgraph.checkpoint.memory import InMemorySaver
from langgraph.graph import END, START, StateGraph
from langgraph.types import Command, interrupt

from ._harness import EventLog, ProbeResult, summarize

LOG = EventLog()


class S(TypedDict, total=False):
    note: str

def _parallel_gate_graph(gate_first: bool):
    def gate(state: S):
        LOG.log("gate:before_interrupt")
        decision = interrupt({"question": "approve the plan?"})
        LOG.log(f"gate:resumed_with={decision}")
        return {"note": f"gate:{decision}"}

    def side_effect(state: S):
        LOG.log("EFFECT:executed")
        return {}

    b = StateGraph(S)
    if gate_first:
        b.add_node("gate", gate)
        b.add_node("side_effect", side_effect)
    else:
        b.add_node("side_effect", side_effect)
        b.add_node("gate", gate)
    b.add_edge(START, "gate")
    b.add_edge(START, "side_effect")
    b.add_edge("gate", END)
    b.add_edge("side_effect", END)
    return b.compile(checkpointer=InMemorySaver())


def probe_sibling_leak(gate_first: bool, label: str) -> ProbeResult:
    LOG.clear()
    g = _parallel_gate_graph(gate_first)
    cfg = {"configurable": {"thread_id": f"t-{label}"}}
    result = g.invoke({"note": ""}, cfg)
    paused = "__interrupt__" in result
    effect_during_pause = LOG.contains("EFFECT:executed")
    trace = list(LOG.events)
    g.invoke(Command(resume=False), cfg)

    return ProbeResult(
        name=f"sibling_leak[{label}]",
        violation=paused and effect_during_pause,
        detail={"paused": paused, "effect_while_paused": effect_during_pause,
                "effect_total_after_reject": LOG.count("EFFECT:executed"),
                "trace_at_pause": trace},
    )

def probe_replay() -> ProbeResult:
    LOG.clear()

    def act_then_gate(state: S):
        LOG.log("P2_EFFECT:executed")
        decision = interrupt({"question": "approve?"})
        LOG.log(f"P2_gate:resumed_with={decision}")
        return {"note": "done"}

    b = StateGraph(S)
    b.add_node("act_then_gate", act_then_gate)
    b.add_edge(START, "act_then_gate")
    b.add_edge("act_then_gate", END)
    g = b.compile(checkpointer=InMemorySaver())
    cfg = {"configurable": {"thread_id": "t-p2"}}
    g.invoke({"note": ""}, cfg)
    at_pause = LOG.count("P2_EFFECT:executed")
    g.invoke(Command(resume=True), cfg)  # APPROVES
    after = LOG.count("P2_EFFECT:executed")

    return ProbeResult(
        name="replay_double_execution",
        violation=after > 1,
        detail={"effect_count_at_pause": at_pause, "effect_count_after_approve": after},
    )

def probe_cancellation(sync_in_thread: bool, label: str) -> ProbeResult:
    LOG.clear()

    def blocking_tool():
        time.sleep(0.6)
        LOG.log(f"{label}_EFFECT:executed_after_delay")

    async def node_sync(state: S):
        LOG.log(f"{label}:node_started")
        await asyncio.to_thread(blocking_tool)
        return {"note": "done"}

    async def node_async(state: S):
        LOG.log(f"{label}:node_started")
        await asyncio.sleep(0.6)
        LOG.log(f"{label}_EFFECT:executed_after_delay")
        return {"note": "done"}

    b = StateGraph(S)
    b.add_node("worker", node_sync if sync_in_thread else node_async)
    b.add_edge(START, "worker")
    b.add_edge("worker", END)
    g = b.compile(checkpointer=InMemorySaver())
    cfg = {"configurable": {"thread_id": f"t-{label}"}}

    async def scenario():
        task = asyncio.create_task(g.ainvoke({"note": ""}, cfg))

        await asyncio.sleep(0.15)
        task.cancel()
        seen = False
        try:
            await task
        except asyncio.CancelledError:
            seen = True
        at = LOG.contains("EFFECT")

        await asyncio.sleep(0.8)
        after = LOG.contains("EFFECT")

        return seen, at, after

    seen, at, after = asyncio.run(scenario())

    return ProbeResult(
        name=f"cancellation[{label}]",
        violation=seen and (not at) and after,
        detail={"cancelled_seen": seen, "effect_at_cancel": at,
                "effect_after_cancel": after},
    )


def probe_timeout_zombie() -> ProbeResult:
    LOG.clear()

    def blocking():
        time.sleep(0.8)
        LOG.log("TO_EFFECT:executed_after_delay")

    async def node(state: S):
        LOG.log("TO:node_started")
        await asyncio.to_thread(blocking)
        return {"note": "done"}

    b = StateGraph(S)
    b.add_node("worker", node)
    b.add_edge(START, "worker")
    b.add_edge("worker", END)
    g = b.compile(checkpointer=InMemorySaver())
    cfg = {"configurable": {"thread_id": "t-to"}}

    async def scenario():
        timed_out = False
        try:
            await asyncio.wait_for(g.ainvoke({"note": ""}, cfg), timeout=0.2)
        except asyncio.TimeoutError:
            timed_out = True
        at = LOG.contains("TO_EFFECT")
        await asyncio.sleep(1.0)
        after = LOG.contains("TO_EFFECT")
        return timed_out, at, after

    timed_out, at, after = asyncio.run(scenario())

    return ProbeResult(
        name="timeout_zombie",
        violation=timed_out and (not at) and after,
        detail={"caller_saw_timeout": timed_out, "effect_at_timeout": at,
                "effect_after_timeout": after},
    )


def main() -> None:
    print(f"# FW-A langgraph=={version('langgraph')}\n")
    results = [
        probe_sibling_leak(True, "gate_first"),
        probe_sibling_leak(False, "effect_first"),
        probe_replay(),
        probe_cancellation(True, "sync_thread"),
        probe_cancellation(False, "pure_async"),
        probe_timeout_zombie(),
    ]
    print(summarize(results))


if __name__ == "__main__":
    main()